### PR TITLE
Add health check, make async Engine more robust

### PR DESCRIPTION
**Pull request by Yard1**
For production usecases, we want to be able to detect Engine failures, especially ones that can happen silently (eg. due to NCCL timeouts). This PR adds a health check method (currently only checking the health of Ray workers) and makes the Async engine more robust by adding a timeout for each iteration as well as better error reporting.

### CODE REVIEW

File: vllm/engine/async_llm_engine.py [hunk 1]
```diff
@@ -14,28 +15,33 @@
...
-        request_tracker.propagate_exception(exc)
-        raise exc
+    except Exception as e:
+        exception = e
+        logger.error("Engine background task failed", exc_info=e)
+        raise AsyncEngineDeadError(
+            msg + " See stack trace above for the actual cause.") from e
+    finally:
+        if exception:
+            error_callback(exception)
```
Comment (zhuohan123) [inline vllm/engine/async_llm_engine.py:44]: We raise errors in both `try` branch and `except` branch. Then what does the `finally` here do?

File: vllm/engine/async_llm_engine.py [hunk 2]
```diff
@@ -161,12 +169,16 @@ def get_new_and_finished_requests(self) -> Tuple[List[Dict], Set[str]]:
...
-
         return new_requests, finished_requests
 
-    async def wait_for_new_requests(self):
-        await self.new_requests_event.wait()
+    async def wait_for_new_requests(self, clear: bool):
+        if not self.has_new_requests():
+            await self.new_requests_event.wait()
+        if clear:
+            self.new_requests_event.clear()
```
Comment (zhuohan123) [inline vllm/engine/async_llm_engine.py:178]: Why don't we always clear this flag?
```suggestion
    async def wait_for_new_requests(self):
        if not self.has_new_requests():
            await self.new_requests_event.wait()
            self.new_requests_event.clear()
```

Comment (zhuohan123) [reply vllm/engine/async_llm_engine.py:178]: Also, what's the reason behind this change? Why do we need to move the `clear` call from `get_new_and_finished_requests` to here?

**Review by zhuohan123 — APPROVED** [review]
Thanks for the contribution! In general LGTM. Left some small questions.

File: vllm/engine/async_llm_engine.py (hunk 1 above)
Comment (Yard1) [reply vllm/engine/async_llm_engine.py:44]: we want to run the error callback even after we re-raise an exception in `except`

File: vllm/engine/async_llm_engine.py (hunk 2 above)
Comment (Yard1) [reply vllm/engine/async_llm_engine.py:178]: Yes, we can always clear it.

The reason is to ensure the event is cleared as soon as we have new requests

File: vllm/engine/async_llm_engine.py [hunk 3]
```diff
@@ -161,12 +169,15 @@ def get_new_and_finished_requests(self) -> Tuple[List[Dict], Set[str]]:
...
 
-        self.new_requests_event.clear()
-
         return new_requests, finished_requests
 
     async def wait_for_new_requests(self):
-        await self.new_requests_event.wait()
+        if not self.has_new_requests():
+            await self.new_requests_event.wait()
+        self.new_requests_event.clear()
```
Comment (njhill) [inline vllm/engine/async_llm_engine.py:177]: Suggestion to only clear before waiting
```suggestion
        if not self.has_new_requests():
            self.new_requests_event.clear()
            if not self.has_new_requests():
                await self.new_requests_event.wait()
```

**Review by njhill — COMMENTED** [review]
Thanks @Yard1 this looks great

File: vllm/engine/async_llm_engine.py (hunk 3 above)
Comment (Yard1) [reply vllm/engine/async_llm_engine.py:177]: Hmm can you explain why we should do it like that?

Comment (Yard1) [reply vllm/engine/async_llm_engine.py:175]: Yeah I think it should be fine

**Pull request merged**